src/core/helpers/bimap.rs

```diff
@@ -0,0 +1,54 @@
+use std::{
+    borrow::Borrow,
+    collections::{btree_map::Range, BTreeMap},
+    ops::RangeBounds,
+    sync::Arc,
+};
+
+pub(crate) struct BiBTreeMap<K, V> {
+    left2right: BTreeMap<Arc<K>, Arc<V>>,
+    right2left: BTreeMap<Arc<V>, Arc<K>>,
+}
+
+impl<K, V> BiBTreeMap<K, V>
+where
+    K: std::hash::Hash + Eq + Ord + std::fmt::Display,
+    V: std::hash::Hash + Eq + Ord,
+{
+    pub(crate) fn new() -> Self {
+        BiBTreeMap {
+            left2right: BTreeMap::new(),
+            right2left: BTreeMap::new(),
+        }
+    }
+
+    pub(crate) fn insert(&mut self, key: K, value: V) {
+        let rc_key = Arc::new(key);
+        let rc_value = Arc::new(value);
+        self.right2left.insert(rc_value.clone(), rc_key.clone());
+        self.left2right.insert(rc_key, rc_value);
+    }
+
+    pub(crate) fn get_by_left(&self, key: &K) -> Option<&Arc<V>> {
+        self.left2right.get(key)
+    }
+
+    pub(crate) fn get_by_right(&self, key: &V) -> Option<&Arc<K>> {
+        self.right2left.get(key)
+    }
+
+    pub(crate) fn range_by_left<T, R>(&self, target: R) -> Range<'_, Arc<K>, Arc<V>>
```

Comment:
Two suggestions, feel free to push back:

- We could implement out own `Range` to avoid exposing `Arc` to the users.
- We could implement `get_by_left` too even though it's not used atm.

Reply:
I am fine with implementing `range_by_left` but isn't implementing our Range too much? If we do not want to expose Arc<> to users another way is to perform here the `next_back()`.. maybe what it makes sense is to implement our own `next_back` for `Arc<K>, Arc<v>`?

This way the user won't need to call `deref`. What do you think?

Reply:
> isn't implementing our Range too much? If we do not want to expose Arc<> to users another way is to perform here the `next_back()`.. maybe what it makes sense is to implement our own `next_back` for `Arc<K>, Arc<v>`?
> 
> This way the user won't need to call `deref`. What do you think?

That's what I had in mind as I don't think implementing `DoubleEndedIterator` (`next_back`) without having your own `Range` is possible. Having said that it should be quite simple: keeping a reference of the map in the custom range and implementing `DoubleEndedIterator` (and `Iterator` maybe). Both only require one function to be implemented, which I think should be a one-liner.

Or did you had something else in mind / am I missing something?